messaginginapp/src/test/java/io/customer/messaginginapp/InAppMessagingStoreTest.kt

```diff
@@ -0,0 +1,372 @@
+
+        val state = manager.getCurrentState()
+        state.messagesInQueue.size shouldBeEqualTo 3
+        (state.currentMessageState as? MessageState.Processing)?.message?.queueId shouldBeEqualTo "2"
+    }
+
+    @Test
+    fun givenDuplicateMessages_whenProcessingMessages_thenDuplicatesAreRemoved() = runTest {
+        val messages = listOf(
+            Message(queueId = "1"),
+            Message(queueId = "1"),
+            Message(queueId = "2")
+        )
+
+        initializeAndSetUser()
+        manager.dispatch(InAppMessagingAction.ProcessMessageQueue(messages))
+
+        val state = manager.getCurrentState()
+        state.messagesInQueue.size shouldBeEqualTo 2
+        state.messagesInQueue.any { it.queueId == "1" } shouldBe true
+        state.messagesInQueue.any { it.queueId == "2" } shouldBe true
+    }
+
+    @Test
+    fun givenMessageWithSpecificRouteRule_whenRouteChanges_thenMessageStateUpdatesAccordingly() = runTest {
+        initializeAndSetUser()
+        val message = Message(queueId = "1", properties = mapOf("gist" to mapOf("routeRuleAndroid" to "home")))
+
+        manager.dispatch(InAppMessagingAction.ProcessMessageQueue(listOf(message)))
+        manager.dispatch(InAppMessagingAction.SetPageRoute("home"))
+
+        var state = manager.getCurrentState()
+        state.currentRoute shouldBe "home"
+        (state.currentMessageState as? MessageState.Processing)?.message?.queueId shouldBe "1"
+
+        manager.dispatch(InAppMessagingAction.SetPageRoute("profile"))
+
+        state = manager.getCurrentState()
+        state.currentRoute shouldBe "profile"
+
+        val currentState = state.currentMessageState
+        currentState shouldBeInstanceOf MessageState.Dismissed::class.java
+        (currentState as MessageState.Dismissed).message.queueId shouldBe "1"
+    }
+
+    @Test
+    fun givenMultipleMessagesWithDifferentRouteRules_whenRouteChanges_thenCorrectMessageIsDisplayed() = runTest {
+        initializeAndSetUser()
+        val homeMessage = Message(queueId = "1", properties = mapOf("gist" to mapOf("routeRuleAndroid" to "home")))
+        val profileMessage = Message(queueId = "1", properties = mapOf("gist" to mapOf("routeRuleAndroid" to "profile")))
+        val generalMessage = Message(queueId = "3")
+
+        // process messages and set initial route
+        manager.dispatch(InAppMessagingAction.ProcessMessageQueue(listOf(homeMessage, profileMessage, generalMessage)))
+        manager.dispatch(InAppMessagingAction.SetPageRoute("home"))
+
+        // verify general message is displayed first (as it has no route rule)
+        var state = manager.getCurrentState()
+        val messageBeingDisplayed = (state.currentMessageState as? MessageState.Processing)?.message
+        messageBeingDisplayed?.queueId shouldBe "3"
+
+        // make the message visible and then dismiss it
+        manager.dispatch(InAppMessagingAction.DisplayMessage(messageBeingDisplayed!!))
```

Comment:
By using `shouldBeInstanceOf`, we should be able to remove all not-null assertion operators in the test